### PR TITLE
Fill data field when api error is not nil on web index

### LINKED ISSUE #5354: Web index does not show error if api call is failed
### What is the issue?
When the web component fails to call Linkerd API the index handler shows a blank page instead of showing an error message.

### How can it be reproduced?
Run the component with the wrong api-addr argument and try to open the index page.

### Logs, error output, etc
```ERRO[2020-12-09T05:09:44+03:30] template: app.tmpl.html:3:33: executing "content" at <.Data.ReleaseVersion>: nil pointer evaluating *public.VersionInfo.ReleaseVersion ```

#### `linkerd check` output

```kubernetes-api
--------------
√ can initialize the client
√ can query the Kubernetes API

kubernetes-version
------------------
√ is running the minimum Kubernetes API version
√ is running the minimum kubectl version

linkerd-existence
-----------------
√ 'linkerd-config' config map exists
√ heartbeat ServiceAccount exist
√ control plane replica sets are ready
√ no unschedulable pods
√ controller pod is running
√ can initialize the client
√ can query the control plane API

linkerd-config
--------------
√ control plane Namespace exists
√ control plane ClusterRoles exist
√ control plane ClusterRoleBindings exist
√ control plane ServiceAccounts exist
√ control plane CustomResourceDefinitions exist
√ control plane MutatingWebhookConfigurations exist
√ control plane ValidatingWebhookConfigurations exist
√ control plane PodSecurityPolicies exist

linkerd-identity
----------------
√ certificate config is valid
√ trust anchors are using supported crypto algorithm
√ trust anchors are within their validity period
√ trust anchors are valid for at least 60 days
√ issuer cert is using supported crypto algorithm
√ issuer cert is within its validity period
‼ issuer cert is valid for at least 60 days
    issuer certificate will expire on 2020-12-10T01:09:48Z
    see https://linkerd.io/checks/#l5d-identity-issuer-cert-not-expiring-soon for hints
√ issuer cert is issued by the trust anchor

linkerd-webhooks-and-apisvc-tls
-------------------------------
√ tap API server has valid cert
√ tap API server cert is valid for at least 60 days
√ proxy-injector webhook has valid cert
√ proxy-injector cert is valid for at least 60 days
√ sp-validator webhook has valid cert
√ sp-validator cert is valid for at least 60 days

linkerd-api
-----------
√ control plane pods are ready
√ control plane self-check
√ [kubernetes] control plane can talk to Kubernetes
√ [prometheus] control plane can talk to Prometheus
√ tap api service is running

linkerd-version
---------------
√ can determine the latest version
√ cli is up-to-date

control-plane-version
---------------------
√ control plane is up-to-date
√ control plane and cli versions match

linkerd-ha-checks
-----------------
√ multiple replicas of control plane pods

Status check results are √
```

### Environment

- Kubernetes Version: 1.16.14
- Cluster Environment: bare-metal
- Host OS: Ubuntu-18.04
- Linkerd version: stable-2.9.0

### Possible solution
The problem can be fixed by filling the Data field of the appParams struct with appropriate values when err is not nil.

### PR DESCRIPTION
**Pull request by keyhana**
app.tmpl.html does not expect the Data field to be empty which only is the case when api call is successful.

This PR fills the Data field with appropriate values when api call returns an error.

Fixes #5354 

Signed-off-by: Keyhan Asadi <keyhana97@gmail.com>

### CODE REVIEW

File: web/srv/handlers.go [hunk 1]
```diff
@@ -55,6 +55,11 @@ func (h *handler) handleIndex(w http.ResponseWriter, req *http.Request, p httpro
 	if err != nil {
 		params.Error = true
 		params.ErrorMessage = err.Error()
+		params.Data = &pb.VersionInfo{
+			GoVersion:      "Unavailable",
+			BuildDate:      "Unavailable",
+			ReleaseVersion: "Unavailable",
```
Comment (cpretzer) [inline web/srv/handlers.go:61]: Is there value in making these strings constants? There may even be a string for this already

File: web/srv/handlers_test.go [hunk 2]
```diff
@@ -64,6 +65,51 @@ func TestHandleIndex(t *testing.T) {
 	}
 }
 
+func TestHandleIndexPublicAPIFail(t *testing.T) {
```
Comment (cpretzer) [inline web/srv/handlers_test.go:68]: Great job adding a test for this!

**Review by cpretzer — COMMENTED** [review]
@keyhana, I cloned your branch and ran the code locally. The change does prevent the blank page from being displayed, and the output in the browser looks like the attached image.

What do you think about making the error messaging more explicit by stating that the `api-addr` is unreachable? Would it make sense to add this as a check in the `linkerd check` output?

Screenshot from 2020-12-08 21-10-36

File: web/srv/handlers.go (hunk 1 above)
Comment (keyhana) [reply web/srv/handlers.go:61]: I did a quick search through the code base and did not find anything useful for this.
I'm not sure if making these consts have a great value but will do so if you think there is.

File: web/srv/handlers_test.go (hunk 2 above)
Comment (keyhana) [reply web/srv/handlers_test.go:68]: Thanks!

**Comment by keyhana** [comment]
@cpretzer I can change the error message to something like this if that's OK.
```Error: <actual error message>, Try to check if api-addr config is set correctly and is reachable.```

I think it would be nice if we also add this as a check.

**Review by adleong — COMMENTED** [review]
Hi @keyhana, thanks for your interest in the project!

I'm not sure that this particular change is valuable since deploying web without the public-api isn't a supported configuration.  It's also not very easy to hit this scenario by accident when using the recommended workflow (e.g. Docker and Kubernetes) so I don't think having a more informative error message for this situation is really that useful.  Even though the change itself is harmless, I'm a bit hesitant to take on code (and especially tests) that don't have a clear purpose.

With all of that said, I'd still be super excited to have you contribute to Linkerd.  I'd recommend looking at [issues tagged "good first issue"](https://github.com/linkerd/linkerd2/issues?q=is%3Aopen+is%3Aissue+label%3A%22good+first+issue%22) or asking in the #contributors channel in the Linkerd slack to find something valuable to work on.

**Comment by keyhana** [comment]
Hello @adleong,
Thanks for your review.
You are right about deploying the web without the public-api, but I think many other things may go wrong in a healthy and complete Linkerd installation, and showing the error message can be helpful in those cases.
Also since the logic for showing the error message is already available in the HTML template file I think the current behavior of not showing it is a bug.
Nevertheless, I will respect your decision on either continuing this PR or closing it.

**Comment by adleong** [comment]
Thanks for understanding, @keyhana, and I hope this doesn't discourage you from contributing.  We'd love to have your participation on the project!

**Pull request closed without merging**